[clang][ExtractAPI] Fix handling of anonymous TagDecls

daniel-grumberg opened the pull request:
This changes the handling of anonymous TagDecls to the following rules:
- If the TagDecl is embedded in the declaration for some VarDecl (this is the only possibility for RecordDecls), then pretend the child decls belong to the VarDecl
- If it's an EnumDecl proceed as we did previously, i.e., embed it in the enclosing DeclContext.

Additionally this fixes a few issues with declaration fragments not consistently including "{ ... }" for anonymous TagDecls. To make testing these additions easier this patch fixes some text declaration fragments merging issues and updates tests accordingly.

daniel-grumberg committed:
[clang][ExtractAPI] Fix handling of anonymous TagDecls

This changes the handling of anonymous TagDecls to the following rules:
- If the TagDecl is embedded in the declaration for some VarDecl (this
  is the only possibility for RecordDecls), then pretend the child decls
  belong to the VarDecl
- If it's an EnumDecl proceed as we did previously, i.e., embed it in
  the enclosing DeclContext.

Additionally this fixes a few issues with declaration fragments not
consistently including "{ ... }" for anonymous TagDecls. To make testing
these additions easier this patch fixes some text declaration fragments
merging issues and updates tests accordingly.

daniel-grumberg commented on clang/lib/ExtractAPI/API.cpp:
```diff
@@ -54,6 +54,13 @@ RecordContext *APIRecord::castToRecordContext(const APIRecord *Record) {
   }
 }
 
+void RecordContext::stealRecordChain(RecordContext &Other) {
+  First = Other.First;
+  Last = Other.Last;
+  Other.First = nullptr;
+  Other.Last = nullptr;
+}
```
Maybe a better name for this would be `transferRecordChain`?

QuietMisdreavus replied:
The name `stealRecordChain` is fine, it makes it plain that `Other`'s record chain will be set to null and moved to this context. I'm just worried with the implementation; if this stands as it is, with no comment and no guard that the current context has no record chain already, then i'm worried about a future use of this function that leads to broken behavior.

If this were a more verbose codebase i would call it `overwriteRecordChainByStealingFrom` but that's a bit too far 😅

daniel-grumberg replied:
There is no harm in "stealing" an empty record chain, and it's not necessarily even semantically wrong, all it indicates is that the current RecordContext won't have any children during any traversals. Semantic problems would still be caught by having a test.

daniel-grumberg committed:
Change semantic of `RecordContext::stealRecordChain` to concatenate the chains

QuietMisdreavus commented on clang/lib/ExtractAPI/API.cpp:
```diff
@@ -54,6 +54,20 @@ RecordContext *APIRecord::castToRecordContext(const APIRecord *Record) {
 
+void RecordContext::stealRecordChain(RecordContext &Other) {
+  // If we don't have an empty chain append Other's chain into ours.
+  if (First)
+    Last->NextInContext = Other.First;
```
This is the tiniest of nitpicks, but the fact that this only checks `First` but then accesses `Last` (which semantically makes sense but is semantically suspect) feels wrong to me. 😅

daniel-grumberg replied:
Documented the invariant that `First` being non-null implies that `Last` is also non-null in an assert above.

QuietMisdreavus reviewed (approved):
Looks good! One last nitpick but you can take it or leave it.

daniel-grumberg committed:
Check invariant of RecordContext

Specifically that the First and Last pointer are always both null or
both non-null.
